### PR TITLE
Split top level of OpenURLServer into Koa middleware PR-1497

### PR DESCRIPTION
**Pull request by skomorokh**

### CODE REVIEW

File: src/OpenURLServer.js [hunk 1]
```diff
@@ -237,8 +101,142 @@ class OpenURLServer {
...
+    Object.keys(serviceConfigs).forEach(label => {
+      this.services[label] = new OkapiSession(cfg, label, serviceConfigs[label]);
+    });
+
+    // Default service
+    if (cfg.getValues().okapiUrl) this.services[''] = new OkapiSession(cfg);
+
+    const docRoot = cfg.getValues().docRoot;
+    if (!docRoot) {
+      throw new Error('No docRoot defined in configuration');
```
Comment (MikeTaylor) [inline src/OpenURLServer.js:197]: Why did you get rid of HTTPError?

**Review by MikeTaylor — COMMENTED** [review]
Unfortunately, this kind of change is very hard to read the diffs for, since a relatively small conceptual change results in a ton of textual change. If you're confident that this is correct, then I'm happy to trust you on that. Hopefully the fairly comprehensive tests will help you to determine that.

File: src/OpenURLServer.js (hunk 1 above)
Comment (skomorokh) [reply src/OpenURLServer.js:197]: Because this is a configuration error not an HTTP error, it happens before the app is even listening for HTTP requests.

Comment (MikeTaylor) [reply src/OpenURLServer.js:197]: D'oh! You're right, of course.

File: src/OpenURLServer.js [hunk 2]
```diff
@@ -1,203 +1,67 @@
...
+  ctx.cfg.log('co', `got ContextObject ${co.getType()} query`, JSON.stringify(co.getQuery()));
+
+  const metadata = co.getMetadata();
+  ctx.cfg.log('metadata', JSON.stringify(metadata));
+
+  ctx.cfg.log('flow', 'Check service');
+  // service can come from path OR parameter
+  const symbol = get(metadata, ['res', 'org']) || ctx.path.replace(/^\//, '');
+  const service = ctx.services[symbol] || ctx.services[''];
+  if (!service) ctx.throw(404, `unsupported service '${symbol}'`);
```
[the remark this answers is not in the document]
Comment (skomorokh) [reply src/OpenURLServer.js:23]: Yes. In both cases it returns the error in the body. I think the other one was also 404 (iirc Koa defaults to that) but it's nice to have it explicit and it's the status that makes most sense here.

https://github.com/openlibraryenvironment/listener-openurl/blob/5ea4d3d5a455661d0da2561680c29e624e050d24/src/OpenURLServer.js#L36

`ctx.throw` is quite handy: https://github.com/koajs/koa/blob/master/docs/api/context.md#ctxthrowstatus-msg-properties

One thing that tripped me up is that it doesn't actually send your error message on a 5xx by default (out of an abundance of caution) unless you tell it to explicitly like so: https://github.com/openlibraryenvironment/listener-openurl/blob/03b39301c940c4f672165e4b339d8a9370327786/src/OpenURLServer.js#L150

**Comment by skomorokh** [comment]
Thanks for looking it over and I definitely appreciated the comprehensive tests. This is exactly the kind of refactoring where you get the most value from good tests and they help me be confident I haven't entirely thrown a wrench in it.

File: src/OpenURLServer.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (MikeTaylor) [reply src/OpenURLServer.js:23]: Sounds good, thanks!

**Comment by MikeTaylor** [comment]
You see, I'm _not_ anti-tests! :-)

**Pull request merged**